### PR TITLE
Allow table name with dots by a PinotConfiguration switch

### PR DESCRIPTION
**Pull request by dongxiaoman**
## Description
Following changes in https://github.com/apache/pinot/pull/5734, 
Allows table name with dots by a shared pinot property. This will allow table name format like `namespace.table_name` to avoid table_name conflict while **providing some mitigation/middle ground before Pinot table goes true multi-tenant**.

The use case this one PR enabled will be for some users sharing a same Pinot Cluster, with the below **Example** rules (not in Pinot, but only with the wrapper/UI or some toolset visible to user):
1. The Admin enforces the rule that the users cannot use dots in their own table_name like `test`;
2. The Admin also enforces the rule that the users will be given a unique "namespace" as prefix to table;
3. The users can create a table named `test` and write query freely/naturally with table name like `user_namespace.test`
4. Pinot can delegate the authorization of table access via its plugin properly in the query

## Extended feature
Extract column name only after the last dot in cases like `SELECT db.namespace.table.column_name from db.namespace.table`. In the example the table name will be `db.namespace.table`

## Tests Done
Many Unit tests

## Release Notes
- Added `pinot.table.name.dots.enabled` config to Controllers and Brokers to allow table name with dots.
- Extract column name from queries like "SELECT db.namespace.table.column_name" properly

### CODE REVIEW

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java [hunk 1]
```diff
@@ -762,23 +763,31 @@ private void handleSubquery(Expression expression, long requestId, JsonNode json
   /**
    * Resolves the actual table name for:
    * - Case-insensitive cluster
-   * - Table name in the format of [database_name].[table_name]
+   * - Table name in the format of [database_name].[namespace].[table_name]
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java:766]: I don't think this change is required. The `[namespace].[table_name]` is the table name, and within pinot there is no concept of `namespace`

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/TableConfigUtils.java [hunk 2]
```diff
@@ -144,10 +145,16 @@ private static Set<ValidationType> parseTypesToSkipString(@Nullable String types
    *   <li>Table name shouldn't contain dot or space in it</li>
    * </ul>
    */
-  public static void validateTableName(TableConfig tableConfig) {
+  public static void validateTableName(TableConfig tableConfig, PinotConfiguration pinotConfiguration) {
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/TableConfigUtils.java:148]: Suggest passing a boolean `allowDot` instead of the whole config

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java [hunk 3]
```diff
@@ -172,6 +172,8 @@ public static class Instance {
     public static final String CONFIG_OF_PINOT_BROKER_STARTABLE_CLASS = "pinot.broker.startable.class";
     public static final String CONFIG_OF_PINOT_SERVER_STARTABLE_CLASS = "pinot.server.startable.class";
     public static final String CONFIG_OF_PINOT_MINION_STARTABLE_CLASS = "pinot.minion.startable.class";
+    public static final String CONFIG_OF_ALLOW_TABLE_NAME_DOTS = "pinot.table.name.dots.enabled";
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java:175]: Let's add it under `Helix` similar to `ENABLE_CASE_INSENSITIVE`. Suggest naming it `allow.table.name.dots`

**Comment by apucher** [comment]
IMO this doesn't even need a feature flag. We're not breaking any behavior. Keeps things simple.

**Comment by Jackie-Jiang** [comment]
@apucher Good point. Let me list down the things we are trying to handle with `.` and discuss a solution:
- `SELECT ... FROM <db>.<table>`
- `SELECT <table>.<col> FROM ...`
In the future, we also want to support:
- `SELECT <col>.<nested_fields> FROM ...`

If we allow `.` within the table name, we cannot tell if `.` is the delimiter or part of the table name.
Add @xiangfu0 to the discussion

**Comment by apucher** [comment]
it should be easy to parse the table name since it needs to be present in the FROM clause.

are nested field selects a thing yet?

**Comment by Jackie-Jiang** [comment]
> it should be easy to parse the table name since it needs to be present in the FROM clause.

I think the `<db>.<table>` convention is from the connector, and we don't have the database concept in Pinot. When getting it from the `FROM` clause, we want to match it with the tables available in pinot. If we support dot in the table name, it will make it slightly hard to parse the name. E.g. when getting `a.b.c.d.e`, we won't know which part is the table name.

> are nested field selects a thing yet?

We don't support it yet, but it is in our roadmap, so we should take them into consideration

I think using dot in table name (or any field name) is in general bad practice because dot is already reserved for other purpose. @dongxiaoman I think we should just use `_` instead of `.` for multi-tenant purpose. To avoid name conflict, we can reject namespace with underlying `_` or perform a prefix check when adding a new namespace.

Some reference about not using `.`:
- https://dba.stackexchange.com/questions/1166/is-it-okay-to-put-dots-in-sql-server-database-names#:~:text=Complex%20SQL%20actions%20are%20processed,if%20the%20base%20is%20fine.&text=These%20errors%20may%20more%20often,of%20anything%20on%20sql%20server.
- https://stackoverflow.com/questions/41948399/is-it-good-practice-to-use-dots-within-table-names-in-mysql

**Comment by apucher** [comment]
There's precedence for using dots as separators across databases/schemas in MySQL, MSSQL, and others. It's tricky in the sense that underscores are much more commonly used in name tables. Both, underscores and dots, could ultimately work. My hunch is that we'll ultimately have to support dots (to some degree) as most cloud analytics systems already do:

* mssql: https://stackoverflow.com/questions/11426929/can-we-use-join-for-two-different-database-tables
(or the links already shared above)
* big query: https://michaelhoweely.com/2020/09/24/how-to-combine-data-from-two-bigquery-public-datasets-using-sql/
* athena: https://stackoverflow.com/questions/58985607/unable-to-join-two-tables-from-two-different-databases-in-athena

**Comment by Jackie-Jiang** [comment]
Agree on using dot to connect db/namespace and table name and column name (this is the standard SQL syntax).
The problem here is that we don't support db/namespace concept natively in Pinot right now, and we are trying to overload the table name to include the namespace as part of it. This can potentially become a blocker for us when adding support of namespace in the future.

**Comment by dongxiaoman** [comment]
> >

@Jackie-Jiang Agree that using dot in table name is not good practice. I already saw JDBC H2 error if I copy/paste the Full integration test because JDBC engine treat dots as separator by default. That is why I hide the support with a feature flag. cc @apucher 
Meanwhile, as I mentioned, table name with dots can provide a transition/middle ground before we fully support full multi-tenant/DB concept.
By this PR, we can allow the user write something naturally:
`SELECT col1 from mydb.test_table where xxx` while the `mydb.test_table` is managed by Pinot admins, and the user is assigned `mydb` from the DataWarehouse infrastructure on top of Pinot.

If this is not breaking other features, providing extra flexibility (extra knob) will help support more use cases, it actually expands Pinot's scope

**Comment by Jackie-Jiang** [comment]
Sounds good. I think we can limit the flexibility of the table name to contain at most one dot, and assume the name is constructed from <db>.<table>. There are 2 places where we handle the dots within the table name:
1. `getActualTableName()`: With the above limitation, no change is needed. The queried table name should have at most one dot
2. `getActualColumnName()`: We can change the logic to do a prefix match for the `rawTableName + '.'` (use `columnName.regionMatches()` for better performance)

File: pinot-segment-local/src/test/java/org/apache/pinot/segment/local/utils/TableConfigUtilsTest.java [hunk 4]
```diff
@@ -860,12 +862,21 @@ public void testTableName() {
...
       try {
-        TableConfigUtils.validateTableName(tableConfig);
+        TableConfigUtils.validateTableName(tableConfig, CommonConstants.Helix.DEFAULT_ALLOW_TABLE_NAME_DOTS);
         Assert.fail("Should fail for malformed table name : " + tableName);
       } catch (IllegalStateException e) {
         // expected
       }
     }
+
+    PinotConfiguration configuration = new PinotConfiguration();
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/test/java/org/apache/pinot/segment/local/utils/TableConfigUtilsTest.java:872]: (minor) Not used

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/TableConfigUtils.java [hunk 5]
```diff
@@ -153,10 +153,13 @@ private static Set<ValidationType> parseTypesToSkipString(@Nullable String types
    *   <li>Table name shouldn't contain dot or space in it</li>
    * </ul>
    */
-  public static void validateTableName(TableConfig tableConfig) {
+  public static void validateTableName(TableConfig tableConfig, boolean allowDots) {
     String tableName = tableConfig.getTableName();
-    if (tableName.contains(".") || tableName.contains(" ")) {
-      throw new IllegalStateException("Table name: '" + tableName + "' containing '.' or space is not allowed");
+    if (tableName.contains(".") && !allowDots) {
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/TableConfigUtils.java:158]: Let's check that the table name contains at most one dot, and put some comments explaining we allow table name with embedded database name

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java [hunk 6]
```diff
@@ -78,6 +78,8 @@ public static class Helix {
     public static final String LEAD_CONTROLLER_RESOURCE_ENABLED_KEY = "RESOURCE_ENABLED";
 
     public static final String ENABLE_CASE_INSENSITIVE_KEY = "enable.case.insensitive";
+    public static final String CONFIG_OF_ALLOW_TABLE_NAME_DOTS = "allow.table.name.dots";
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java:81]: Per the discussion, suggest changing it to `ALLOW_TABLE_NAME_WITH_DATABASE`, and only allow one dot if enabled

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java [hunk 7]
```diff
@@ -765,31 +766,41 @@ private void handleSubquery(Expression expression, long requestId, JsonNode json
...
+   * If the PinotConfiguration allows dots in table name, we will not do the split
+   * @param tableName the table name in the query
+   * @param tableCache the table case-sensitive cache
+   * @param routingManager the routing mananger for testing whether a route exists
+   * @param config the Pinot Configuration
+   * @return
    */
-  private String getActualTableName(String tableName) {
+  @VisibleForTesting
+  static String getActualTableName(String tableName, TableCache tableCache, BrokerRoutingManager routingManager,
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java:779]: We might not want to change this method. Even if we allow table name with database, we should probably still support the case of table name without database.

Comment (dongxiaoman) [reply pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java:776]: Right now the code change has only one effect:
If we allow dots in table name, we will avoid striping the `[database_name].` from the table name. This will be needed in our case, that we wish to treat `[database_name].[table_name]` as the whole actual table name.

This behavior is under the feature flag so will be backwards compatible.

File: pinot-common/src/main/java/org/apache/pinot/common/config/provider/TableCache.java [hunk 8]
```diff
@@ -76,7 +76,7 @@ public class TableCache implements PinotConfigProvider {
   private final Set<SchemaChangeListener> _schemaChangeListeners = new HashSet<>();
 
   private final ZkHelixPropertyStore<ZNRecord> _propertyStore;
-  private final boolean _caseInsensitive;
+  private final boolean _ignoreCase;
```
Comment (dongxiaoman) [inline pinot-common/src/main/java/org/apache/pinot/common/config/provider/TableCache.java:79]: @Jackie-Jiang  I renamed the `caseInSensitive` everywhere to `ignoreCase` because I already got confused during my coding. Negating in naming will cause more trouble.

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java (hunk 6 above)
Comment (dongxiaoman) [reply pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java]: Changed.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/TableConfigUtils.java (hunk 5 above)
Comment (dongxiaoman) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/TableConfigUtils.java]: Added logic and also unit tests

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java (hunk 7 above)
Comment (Jackie-Jiang) [reply pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java:776]: What I meant is that the existing code should already be able to handle the `[database_name].[table_name]` as the table name (line 773, 801, 812 in the original code). We don't need to introduce the extra logic of reading properties from the config.

Comment (dongxiaoman) [reply pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java]: I see. I did not expect the behavior to be the same but you are right.
I am still keeping the unit tests since they will ensure our expected behavior is not broken

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java [hunk 9]
```diff
@@ -762,29 +763,34 @@ private void handleSubquery(Expression expression, long requestId, JsonNode json
    * - Case-insensitive cluster
    * - Table name in the format of [database_name].[table_name]
    *
-   * Drop the database part if there is no existing table in the format of [database_name].[table_name], but only
-   * [table_name].
+   * If the PinotConfiguration does not allow dots in table name, will do the below:
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java:766]: Let's revert the comment update since we are following the old behavior

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java [hunk 10]
```diff
@@ -762,29 +763,34 @@ private void handleSubquery(Expression expression, long requestId, JsonNode json
...
+   * @param routingManager the routing mananger for testing whether a route exists
+   * @return the table name in the format of [database_name].[table_name]
    */
-  private String getActualTableName(String tableName) {
+  @VisibleForTesting
+  static String getActualTableName(String tableName, TableCache tableCache, BrokerRoutingManager routingManager) {
     // Use TableCache to handle case-insensitive table name
-    if (_tableCache.isCaseInsensitive()) {
-      String actualTableName = _tableCache.getActualTableName(tableName);
+    if (tableCache.isIgnoreCase()) {
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java:777]: I just realize that we have enhanced `tableCache` in #8475 to store table name regardless of the case sensitivity. We can actually significantly simplify this method by always using `tableCache` to resolve the actual table name:
```suggestion
static String getActualTableName(String tableName, TableCache tableCache) {
    String actualTableName = tableCache.getActualTableName(tableName);
    if (actualTableName != null) {
      return actualTableName;
    }
    
    // Check if table is in the format of [database_name].[table_name]
    String[] tableNameSplits = StringUtils.split(tableName, ".", 2);
    if (tableNameSplits.length == 2) {
      actualTableName = tableCache.getActualTableName(tableNameSplits[1]);
      if (actualTableName != null) {
        return actualTableName;
      }
    }

    return tableName;
  }
```

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java [hunk 11]
```diff
@@ -183,7 +184,9 @@ public ConfigSuccessResponse addTable(
       TableConfigUtils.validate(tableConfig, schema, typesToSkip, _controllerConf.isDisableIngestionGroovy());
       // TableConfigUtils.validateTableName(...) checks table name rules.
       // So it won't effect already created tables.
-      TableConfigUtils.validateTableName(tableConfig);
+      boolean allowDots = _controllerConf.getProperty(CommonConstants.Helix.ALLOW_TABLE_NAME_WITH_DATABASE,
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java:187]: (minor) Rename to `allowTableNameWithDatabase`

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/TableConfigsRestletResource.java [hunk 12]
```diff
@@ -408,19 +409,20 @@ private TableConfigs validateConfig(TableConfigs tableConfigs, @Nullable String
       Preconditions.checkState(rawTableName.equals(schema.getSchemaName()),
           "'tableName': %s must be equal to 'schemaName' from 'schema': %s", rawTableName, schema.getSchemaName());
       SchemaUtils.validate(schema);
-
+      boolean allowDots = _controllerConf.getProperty(CommonConstants.Helix.ALLOW_TABLE_NAME_WITH_DATABASE,
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/TableConfigsRestletResource.java:412]: (minor) Rename to `allowTableNameWithDatabase`

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/TableConfigUtils.java [hunk 13]
```diff
@@ -153,10 +153,19 @@ private static Set<ValidationType> parseTypesToSkipString(@Nullable String types
    *   <li>Table name shouldn't contain dot or space in it</li>
    * </ul>
    */
-  public static void validateTableName(TableConfig tableConfig) {
+  public static void validateTableName(TableConfig tableConfig, boolean allowDots) {
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/TableConfigUtils.java:156]: (minor) Rename to `allowTableNameWithDatabase`

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/TableConfigUtils.java [hunk 14]
```diff
@@ -153,10 +153,19 @@ private static Set<ValidationType> parseTypesToSkipString(@Nullable String types
    *   <li>Table name shouldn't contain dot or space in it</li>
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/TableConfigUtils.java:153]: Update the javadoc about the new validation logic

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/TableConfigUtils.java [hunk 15]
```diff
@@ -153,10 +153,19 @@ private static Set<ValidationType> parseTypesToSkipString(@Nullable String types
...
+    int dotCount = StringUtils.countMatches(tableName, '.');
+    // For transitioning into full [database_name].[table_name] support, we allow the table name
+    // with one dot at max, so the admin may create mydb.mytable with a feature knob.
+    if (allowDots && dotCount > 1) {
+      throw new IllegalStateException("Table name: '" + tableName + "' containing more than one '.' is not allowed");
+    }
+    if (!allowDots && dotCount > 0) {
+      throw new IllegalStateException("Table name: '" + tableName + "' containing '.' is not allowed");
+    }
+    if (tableName.contains(" ")) {
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/TableConfigUtils.java:167]: Not introduced in this PR, but more robust
```suggestion
    if (StringUtils.containsWhitespace(tableName)) {
```

**Review by Jackie-Jiang — APPROVED** [review]
LGTM other wise

**Comment by dongxiaoman** [comment]
@Jackie-Jiang Sorry for the late response. Can you please have another quick look? All comments addressed
